client/src/components/logViewer/logParser.ts

```diff
@@ -0,0 +1,438 @@
+// Copyright © 2023, SAS Institute Inc., Cary, NC, USA.  All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// It can't be shown from the definition, but actually LogType can be uppercase, partial uppercase, as well as lowercase
+// for example, 'ERROR' and 'Error' are acceptable and will be treated as 'error'
+type LogType = "error" | "warning" | "note" | "title" | "source" | "normal";
+
+type ProblemType = "error" | "warning";
+
+export type LogItem = {
+  line: string;
+  type: LogType;
+  [key: string]: number | string;
+};
+
+export type Problem = {
+  lineNumber: number;
+  startColumn: number;
+  endColumn: number;
+  message: string;
+  type: ProblemType;
+};
+
+type ProblemLocation = {
+  startColumn: number;
+  endColumn: number;
+  lineNumber: number;
+};
+
+// the start column for code in log may be different with which is in source file,
+// one case is the \t is converted to multiple spaces, so column offset is needed.
+// the line number is different with which is in source file as well, so line offset is needed.
+type LocationOffset = {
+  columnOffset: number;
+  lineOffset: number;
+};
+
+type ProblemMetadata = {
+  codeLogLine: string;
+  logLines: string[];
+  location: ProblemLocation;
+  // "   -----" {location indicator}
+  // "   22" {problem index at same location}
+  // "   76" {problem index at same location}
+  problemIndicsForSameLocation: string[];
+  type: ProblemType; //"error" | "warning";
+};
+
+export function parseLog(
+  logs: LogItem[],
+  firstCodeLine: string,
+): Problem[] | null {
+  if (logs.length === 0 || firstCodeLine.trim() === "") {
+    return null;
+  }
+
+  const latestLogs = filterOutTheLatestLogs(logs, firstCodeLine);
+
+  const problemRelatedLogs = filterOutProblemRelatedLogs(latestLogs);
+
+  const problems: Problem[] = [];
```

Comment:
Can we move this variable declaration to _just_ before the `.forEach` on line 72. I think that'd make it a little more clear where we used `problems`

Reply:
It makes sense. Done